pkg/certificate/auto/provider.go

```diff
@@ -0,0 +1,295 @@
+package auto
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"time"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	interfaces "go.etcd.io/etcd-operator/pkg/certificate/interfaces"
+	"go.etcd.io/etcd/client/pkg/v3/transport"
+)
+
+const (
+	DefaultValidity = 365 * 24 * time.Hour
+)
+
+type AutoCertProvider struct {
+	client.Client
+}
+
+var _ interfaces.Provider = (*AutoCertProvider)(nil)
+
+func New(c client.Client) interfaces.Provider {
+	return &AutoCertProvider{
+		c,
+	}
+}
+
+func (ac *AutoCertProvider) EnsureCertificateSecret(ctx context.Context, secretName, namespace string,
+	cfg *interfaces.Config) error {
+	var secret corev1.Secret
+	err := ac.Get(ctx, client.ObjectKey{Name: secretName, Namespace: namespace}, &secret)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			err := ac.createNewSecret(ctx, secretName, namespace, cfg)
+			if err != nil {
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+
+	err = ac.ValidateCertificateSecret(ctx, secretName, namespace, cfg)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return err
+		} else {
+			return fmt.Errorf("invalid certificate secret: %s present in namespace: %s, please delete and try again.\nError: %s",
```

Comment:
Suggestion to wrap the underlying error:

```suggestion
			return fmt.Errorf("invalid certificate secret: %s present in namespace: %s, please delete and try again.\nError: %w",
```